File: lib/src/test/java/uk/co/glass_software/android/shared_preferences/persistence/base/StoreEntryUnitTest.java
/*
 * Copyright (C) 2017 Glass Software Ltd
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package uk.co.glass_software.android.shared_preferences.persistence.base;


import org.junit.Before;
import org.junit.Test;

import uk.co.glass_software.android.shared_preferences.StoreKey;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class StoreEntryUnitTest {
    
    private StoreEntry<String> target;
    private KeyValueStore mockStore;
    private StoreKey storeKey = StoreKey.TEST;
    
    private class TestEntry extends StoreEntry<String> {
        private TestEntry(KeyValueStore store) {
            super(store, storeKey, storeKey);
        }
    }
    
    @Before
    public void setUp() throws Exception {
        mockStore = mock(KeyValueStore.class);
        target = new TestEntry(mockStore);
    }
    
    @Test
    public void testSaveNull() {
        target.save(null);
        verify(mockStore).saveValue(eq(storeKey.getUniqueKey()), eq(null));
    }
    
    @Test
    public void testSave() {
        String value = "some string";
        target.save(value);
        verify(mockStore).saveValue(eq(storeKey.getUniqueKey()), eq(value));
    }
    
    @Test
    public void testGet() {
        String test = "test";
        when(mockStore.getValue(eq(storeKey.getUniqueKey()), eq(String.class), eq(null))).thenReturn(test);
        
        String result = target.get();
        
        verify(mockStore).getValue(eq(storeKey.getUniqueKey()), eq(String.class), eq(null));
        assertEquals("Strings don't match", test, result);
    }
    
    @Test
    public void testGetWithDefault() {
        String string = "abc";
        String defaultValue = "default";
        when(mockStore.getValue(eq(storeKey.getUniqueKey()), eq(String.class), eq(defaultValue)))
                .thenReturn(string);
        
        String result = target.get(defaultValue);
        
        verify(mockStore).getValue(eq(storeKey.getUniqueKey()), eq(String.class), eq(defaultValue));
        assertEquals("Strings don't match", string, result);
    }
    
    @Test
    public void testDrop() {
        target.drop();
        verify(mockStore).deleteValue(eq(storeKey.getUniqueKey()));
    }
    
    @Test
    public void testGetKey() {
        String key = target.getKey();
        assertEquals("Key is wrong", storeKey.getUniqueKey(), key);
    }
    
    @Test
    public void testGetKeyString() {
        String key = target.getKey();
        assertEquals("Key is wrong", storeKey.getUniqueKey(), key);
    }
    
    @Test
    public void testExists() throws Exception {
        when(mockStore.hasValue(eq(storeKey.getUniqueKey()))).thenReturn(true);
        assertTrue("Entry should exists", target.exists());

        when(mockStore.hasValue(eq(storeKey.getUniqueKey()))).thenReturn(false);
        assertFalse("Entry should exists", target.exists());

        verify(mockStore, times(2)).hasValue(eq(storeKey.getUniqueKey()));
    }
}

File: lib/src/test/java/uk/co/glass_software/android/shared_preferences/persistence/base/StoreEntryUnitTest.kt
/*
 * Copyright (C) 2017 Glass Software Ltd
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package uk.co.glass_software.android.shared_preferences.persistence.base

import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.isNull
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import junit.framework.Assert.*
import org.junit.Before
import org.junit.Test
import uk.co.glass_software.android.shared_preferences.persistence.preferences.StoreEntry
import uk.co.glass_software.android.shared_preferences.test.verifyWithContext

class StoreEntryUnitTest {


    private lateinit var target: KeyValueEntry<String>
    private lateinit var mockStore: KeyValueStore

    private inner class TestEntry(store: KeyValueStore) : StoreEntry<String>(
            store,
            KEY,
            String::class.java
    )

    @Before
    @Throws(Exception::class)
    fun setUp() {
        mockStore = mock()
        target = TestEntry(mockStore)
    }

    @Test
    fun testSaveNull() {
        target.save(null)

        verifyWithContext(mockStore).saveValue<Any>(
                eq(KEY),
                isNull()
        )
    }

    @Test
    fun testSave() {
        val value = "some string"

        target.save(value)

        verifyWithContext(mockStore).saveValue(
                eq(KEY),
                eq(value)
        )
    }

    @Test
    fun testGet() {
        val test = "test"

        whenever(mockStore.getValue(
                eq(KEY),
                eq(String::class.java)
        )).thenReturn(test)

        val result = target.get()

        verifyWithContext(mockStore).getValue(
                eq(KEY),
                eq(String::class.java)
        )

        assertEquals("Strings don't match", test, result)
    }

    @Test
    fun testGetWithDefault() {
        val string = "abc"
        val defaultValue = "default"

        whenever(mockStore.getValue(
                eq(KEY),
                eq(String::class.java),
                eq(defaultValue)
        )).thenReturn(string)

        val result = target.get(defaultValue)

        verifyWithContext(mockStore).getValue(
                eq(KEY),
                eq(String::class.java),
                eq(defaultValue)
        )

        assertEquals("Strings don't match", string, result)
    }

    @Test
    fun testDrop() {
        target.drop()

        verifyWithContext(mockStore)
                .deleteValue(eq(KEY))
    }

    @Test
    fun testGetKey() {
        val key = target.getKey()

        assertEquals("Key is wrong", KEY, key)
    }

    @Test
    @Throws(Exception::class)
    fun testExists() {
        whenever(
                mockStore.hasValue(eq(KEY))
        ).thenReturn(true)

        assertTrue("Entry should exists", target.exists())

        whenever(
                mockStore.hasValue(eq(KEY))
        ).thenReturn(false)

        assertFalse("Entry should exists", target.exists())

        verifyWithContext(mockStore).hasValue(eq(KEY))
    }

    companion object {
        private const val KEY = "KEY"
    }
}